components/net/image_cache.rs

```diff
@@ -559,11 +527,20 @@ impl ImageCacheStore {
 
     /// Insert received keys into the cache and complete the loading of images
     fn insert_keys_and_load_images(&mut self, image_keys: Vec<WebRenderImageKey>) {
-        for image_key in image_keys {
-            if let Some(pending_key) = self.key_cache.no_key_yet.pop_front() {
-                self.set_key_and_finish_load(pending_key, image_key);
-            } else {
-                self.key_cache.add_key(image_key);
+        if let KeyCacheState::Ready(ref mut queue) = self.key_cache.cache {
+            for key in image_keys {
+                queue.push_back(key);
+            }
+        } else {
+            let len = self.key_cache.images_pending_keys.len();
+            self.key_cache.cache = KeyCacheState::Ready(image_keys.into());
+            let images = self
+                .key_cache
+                .images_pending_keys
+                .drain(0..len)
+                .collect::<Vec<PendingKey>>();
```

Comment:
This is correct but unnecessarily puts back images into `images_pending_keys`, so I would say it's cleaner to make `len` be the smaller of  `image_keys.len()` and `images_pending_keys.len()`.

Reply:
So here I was wrong, because only processing what I defined as `len` leads to the same livelock I had identified earlier(a state where `cache.is_empty()` is true for `Ready(cache)`, but there are still images pending keys).

So after the `for key in images` loop, we need to check the len of the cache, and if it is empty, request a new batch and set the state back to `PendingBatch`.